packages/block-editor/src/hooks/color-panel.js

```diff
@@ -29,32 +30,22 @@ export default function ColorPanel( {
 	const definedColors = settings.filter( ( setting ) => setting?.colorValue );
 
 	useEffect( () => {
+		// If contrast checking is disabled, don't do anything.
 		if ( ! enableContrastChecking ) {
 			return;
 		}
-		if ( ! definedColors.length ) {
-			if ( detectedBackgroundColor ) {
-				setDetectedBackgroundColor();
-			}
-			if ( detectedColor ) {
-				setDetectedColor();
-			}
-			if ( detectedLinkColor ) {
-				setDetectedColor();
-			}
-			return;
-		}
 
-		if ( ! ref.current ) {
-			return;
-		}
+		// Try to detect the colors for the text, link and backrgound based on the
+		// current ref.
 		setDetectedColor( getComputedStyle( ref.current ).color );
 
 		const firstLinkElement = ref.current?.querySelector( 'a' );
 		if ( firstLinkElement && !! firstLinkElement.innerText ) {
 			setDetectedLinkColor( getComputedStyle( firstLinkElement ).color );
 		}
 
+		// If the current node has a transparent background color, we need to
+		// find the first parent node with a non-transparent background color.
```

Comment:
Nit: The first two lines below this comment appear more related to the previous inline comment about color detection. Could we move this comment to immediately above the while loop that it actually relates to?